lib/maths/CBoostedTreeFactory.cc

```diff
@@ -0,0 +1,340 @@
+
+    TPackedBitVectorVec trainingRowMasks(m_Tree->m_Impl->m_NumberFolds);
+
+    for (auto row = mask.beginOneBits(); row != mask.endOneBits(); ++row) {
+        std::size_t fold{CSampling::uniformSample(m_Tree->m_Impl->m_Rng, 0,
+                                                  m_Tree->m_Impl->m_NumberFolds)};
+        trainingRowMasks[fold].extend(true, *row - trainingRowMasks[fold].size());
+        trainingRowMasks[fold].extend(false);
+    }
+
+    for (auto& fold : trainingRowMasks) {
+        fold.extend(true, mask.size() - fold.size());
+        LOG_TRACE(<< "# training = " << fold.manhattan());
+    }
+
+    TPackedBitVectorVec testingRowMasks(m_Tree->m_Impl->m_NumberFolds,
+                                        core::CPackedBitVector{mask.size(), true});
+    for (std::size_t i = 0; i < m_Tree->m_Impl->m_NumberFolds; ++i) {
+        testingRowMasks[i] ^= trainingRowMasks[i];
+        LOG_TRACE(<< "# testing = " << testingRowMasks[i].manhattan());
+    }
+
+    return {trainingRowMasks, testingRowMasks};
+}
+
+//! Initialize the regressors sample distribution.
+void CBoostedTreeFactory::initializeFeatureSampleDistribution(const core::CDataFrame& frame) {
+
+    // Exclude all constant features by zeroing their probabilities.
+
+    std::size_t n{m_Tree->m_Impl->numberFeatures(frame)};
+
+    TSizeVec regressors(n);
+    std::iota(regressors.begin(), regressors.end(), 0);
+    regressors.erase(regressors.begin() + m_Tree->m_Impl->m_DependentVariable);
+
+    TDoubleVecVec distinct(n);
+    frame.readRows(1, [&](TRowItr beginRows, TRowItr endRows) {
+        for (auto row = beginRows; row != endRows; ++row) {
+            for (std::size_t i = 0; i < regressors.size(); ++i) {
+                double value{(*row)[regressors[i]]};
+                if (distinct[i].size() == 2) {
+                    continue;
+                }
+                if (distinct[i].empty()) {
+                    distinct[i].push_back(value);
+                }
+                if (value != distinct[i][0]) {
+                    distinct[i].push_back(value);
+                }
+            }
+        }
+    });
+
+    regressors.erase(
+        std::remove_if(regressors.begin(), regressors.end(),
+                       [&](std::size_t i) { return distinct[i].size() < 2; }),
+        regressors.end());
+    LOG_TRACE(<< "regressors = " << core::CContainerPrinter::print(regressors));
+
+    // TODO consider "correlation" with target variable.
+
+    m_Tree->m_Impl->m_FeatureSampleProbabilities.assign(n, 0.0);
+    if (regressors.empty()) {
+        HANDLE_FATAL(<< "Input error: all features constant.");
+    } else {
+        double p{1.0 / static_cast<double>(regressors.size())};
+        for (auto feature : regressors) {
+            m_Tree->m_Impl->m_FeatureSampleProbabilities[feature] = p;
+        }
+    }
+    LOG_TRACE(<< "P(sample) = "
+              << core::CContainerPrinter::print(m_Tree->m_Impl->m_FeatureSampleProbabilities));
```

Comment:
Just spotted that this is out-of-date: predates #488. I'm thinking it might be better to reapply to a clean branch of feature/regression (if it isn't too much trouble).

Reply:
I rebased in and created PR #555